In [78]:
import pandas as pd
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report

# Exercises I

### 1. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

#### 1. 1. In the context of this problem, what is a false positive?

There are 13 False Positives, where we predicted dog but the result was cat.

#### 1. 2. In the context of this problem, what is a false negative?

There are 7 False Negatives, where we predicted cat but the result was dog.

#### 1. 3. How would you describe this model?

In [2]:
acc_dog = round((46+36)/(46+36+7+13)*100 ,2)
print(f"The model's overall accuracy of predicting dog correctly is {acc_dog}%")

The model's overall accuracy of predicting dog correctly is 80.39%


I'd say it's a pretty good model, accuracy is pretty viable but I'd love to see much more samples to determine a better overall accuracy.

### 2. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant. 
### Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

In [3]:
C3 = pd.read_csv('/Users/jay/codeup-data-science/Classification/classification-exercises/c3.csv')
C3

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

#### 2. 1. Use the predictions dataset and pandas to help answer the following questions:

##### 2. 1. 1. An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

The recall evaluation will give us the best representation of how well the models captured the defects, model 3 had the highest percentage at 81.25%.

In [106]:
C3['baseline_prediction'] = 'Defect'

In [5]:
C3['actual'].value_counts()

actual
No Defect    184
Defect        16
Name: count, dtype: int64

In [6]:
model_1 = pd.crosstab(C3['actual'], C3['model1']).T
model_1

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [7]:
(8)/(8+8)

0.5

In [9]:
model_2 = pd.crosstab(C3['actual'], C3['model2']).T
model_2

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [10]:
(9)/(9+7)

0.5625

In [12]:
model_3 = pd.crosstab(C3['actual'], C3['model3']).T
model_3

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [14]:
(13)/(3+13)

0.8125

In [111]:
subset = C3[C3['actual'] == 'Defect']
model_recall1 = (subset['model1'] == subset['actual']).mean()
model_recall2 = (subset['model2'] == subset['actual']).mean()
model_recall3 = (subset['model3'] == subset['actual']).mean()
print(f'  model recall 1: {model_recall1:.2%}')
print(f'  model recall 2: {model_recall2:.2%}')
print(f'  model recall 3: {model_recall3:.2%}')

  model recall 1: 50.00%
  model recall 2: 56.25%
  model recall 3: 81.25%


##### 2. 1. 2. Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

A precision evaluation was needed since we are trying to find all Defects whether actual or False Positive. Model 3 had the best model precision at 97.03% to the baseline 92%.

In [16]:
model_1

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [17]:
subset = C3[C3['model1'] == 'Defect']
model_precision = (subset.model1 == subset.actual).mean()

subset = C3[C3.baseline_prediction == 'Defect']
baseline_precision = (subset.baseline_prediction == subset.actual).mean()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

model precision: 95.79%
baseline precision: 92.00%


In [18]:
model_2

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [19]:
subset = C3[C3['model2'] == 'No Defect']
model_precision = (subset.model2 == subset.actual).mean()

subset = C3[C3.baseline_prediction == 'No Defect']
baseline_precision = (subset.baseline_prediction == subset.actual).mean()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

model precision: 93.64%
baseline precision: 92.00%


In [20]:
model_3

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [113]:
subset1 = C3[C3['model1'] == 'Defect']
model_precision1 = (subset.model1 == subset.actual).mean()
subset2 = C3[C3['model2'] == 'Defect']
model_precision2 = (subset.model2 == subset.actual).mean()
subset3 = C3[C3['model3'] == 'Defect']
model_precision3 = (subset.model3 == subset.actual).mean()

print(f'model precision 1: {model_precision1:.2%}')
print(f'model precision 2: {model_precision2:.2%}')
print(f'model precision 3: {model_precision3:.2%}')

model precision 1: 50.00%
model precision 2: 56.25%
model precision 3: 81.25%


### 3. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee). At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

#### 3. 1. Several models have already been developed with the data, and you can find their results here.

In [22]:
paws = pd.read_csv('/Users/jay/codeup-data-science/Classification/classification-exercises/gives_you_paws.csv')
paws

actual model1 model2 model3 model4
0       cat    cat    dog    cat    dog
1       dog    dog    cat    cat    dog
2       dog    cat    cat    cat    dog
3       dog    dog    dog    cat    dog
4       cat    cat    cat    dog    dog
...     ...    ...    ...    ...    ...
4995    dog    dog    dog    dog    dog
4996    dog    dog    cat    cat    dog
4997    dog    cat    cat    dog    dog
4998    cat    cat    cat    cat    dog
4999    dog    dog    dog    dog    dog

[5000 rows x 5 columns]

#### 3. 2. Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [23]:
paws['actual'].value_counts()

actual
dog    3254
cat    1746
Name: count, dtype: int64

In [24]:
paws['baseline_prediction'] = 'dog'

In [25]:
baseline_accuracy = (paws.baseline_prediction == paws.actual).mean()
baseline_accuracy

0.6508

##### 3. 2. 1. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In terms of accuracy, Model 1 performs the best at 80.74% to the baseline accuracy of 65.08%. Model 4 also beat baseline at 74% to baseline 65%. Model 2 and 3 perform worse than baseline.

In [30]:
model_1 = pd.crosstab(paws['actual'], paws['model1']).T
model_1

actual   cat   dog
model1            
cat     1423   640
dog      323  2614

In [32]:
(1423+2614)/(1423+2614+640+323)

0.8074

In [31]:
model_2 = pd.crosstab(paws['actual'], paws['model2']).T
model_2

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [33]:
(1555+1597)/(1555+1597+1657+191)

0.6304

In [29]:
model_3 = pd.crosstab(paws['actual'], paws['model3']).T
model_3

actual  cat   dog
model3           
cat     893  1599
dog     853  1655

In [34]:
(893+1655)/(893+1599+853+1655)

0.5096

In [118]:
baseline_accuracy = (paws.baseline_prediction == paws.actual).mean()
model_accuracy1 = (paws.model1 == paws.actual).mean()
model_accuracy2 = (paws.model2 == paws.actual).mean()
model_accuracy3 = (paws.model3 == paws.actual).mean()
model_accuracy4 = (paws.model4 == paws.actual).mean()

print(f'baseline accuracy: {baseline_accuracy:.2%}')
print(f' model 1 accuracy: {model_accuracy1:.2%}')
print(f' model 2 accuracy: {model_accuracy2:.2%}')
print(f' model 3 accuracy: {model_accuracy3:.2%}')
print(f' model 4 accuracy: {model_accuracy4:.2%}')

baseline accuracy: 65.08%
 model 1 accuracy: 80.74%
 model 2 accuracy: 63.04%
 model 3 accuracy: 50.96%
 model 4 accuracy: 74.26%


##### 3. 2. 2. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?

Model 4 has the highest precision accuracy for predicting dogs at almost 96%.

In [38]:
model_1

actual   cat   dog
model1            
cat     1423   640
dog      323  2614

In [54]:
(2614)/(2614+323)

0.8900238338440586

In [39]:
model_2

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [55]:
(1597)/(1597+191)

0.8931767337807607

In [40]:
model_3

actual  cat   dog
model3           
cat     893  1599
dog     853  1655

In [56]:
(1655)/(1655+853)

0.6598883572567783

In [132]:
paws['baseline_prediction'] = 'dog'
subset = paws[paws['actual'] == 'dog']
baseline_accuracy = (paws.baseline_prediction == paws.actual).mean()
model_recall1 = (subset['model1'] == subset['actual']).mean()
model_recall2 = (subset['model2'] == subset['actual']).mean()
model_recall3 = (subset['model3'] == subset['actual']).mean()
model_recall4 = (subset['model4'] == subset['actual']).mean()
print(f'   model 1 recall: {model_recall1:.2%}')
print(f'   model 2 recall: {model_recall2:.2%}')
print(f'   model 3 recall: {model_recall3:.2%}')
print(f'   model 4 recall: {model_recall4:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model 1 recall: 80.33%
   model 2 recall: 49.08%
   model 3 recall: 50.86%
   model 4 recall: 95.57%
baseline accuracy: 65.08%


##### 3. 2. 3. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?

Model 2 has the best recall percentage at 89.06%. 

Model 1 is close behind it at 81.50%.

In [41]:
model_1

actual   cat   dog
model1            
cat     1423   640
dog      323  2614

In [47]:
(1423)/(1423+640)

0.6897721764420747

In [42]:
model_2

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [46]:
(1555)/(1555+1657)

0.4841220423412204

In [43]:
model_3

actual  cat   dog
model3           
cat     893  1599
dog     853  1655

In [45]:
(893)/(893+1599)

0.358346709470305

In [133]:
paws['baseline_prediction'] = 'cat'
subset = paws[paws['actual'] == 'cat']
baseline_accuracy = (paws.baseline_prediction == paws.actual).mean()
model_recall1 = (subset['model1'] == subset['actual']).mean()
model_recall2 = (subset['model2'] == subset['actual']).mean()
model_recall3 = (subset['model3'] == subset['actual']).mean()
model_recall4 = (subset['model4'] == subset['actual']).mean()
print(f'   model 1 recall: {model_recall1:.2%}')
print(f'   model 2 recall: {model_recall2:.2%}')
print(f'   model 3 recall: {model_recall3:.2%}')
print(f'   model 4 recall: {model_recall4:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model 1 recall: 81.50%
   model 2 recall: 89.06%
   model 3 recall: 51.15%
   model 4 recall: 34.54%
baseline accuracy: 34.92%


### 4. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

#### 4. 1. sklearn.metrics.accuracy_score

In [160]:
accuracy_score?

Signature: accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None)
Docstring:
Accuracy classification score.

In multilabel classification, this function computes subset accuracy:
the set of labels predicted for a sample must *exactly* match the
corresponding set of labels in y_true.

Read more in the :ref:`User Guide <accuracy_score>`.

Parameters
----------
y_true : 1d array-like, or label indicator array / sparse matrix
    Ground truth (correct) labels.

y_pred : 1d array-like, or label indicator array / sparse matrix
    Predicted labels, as returned by a classifier.

normalize : bool, default=True
    If ``False``, return the number of correctly classified samples.
    Otherwise, return the fraction of correctly classified samples.

sample_weight : array-like of shape (n_samples,), default=None
    Sample weights.

Returns
-------
score : float
    If ``normalize == True``, return the fraction of correctly
    classified samples (float), else returns the number of c

In [188]:
acc_mod_1 = accuracy_score(paws['model1'],paws['actual'])
print(f'The accuracy score for Model 1 is {acc_mod_1:.2%}')

The accuracy score for Model 1 is 80.74%


In [189]:
acc_mod_2 = accuracy_score(paws['model2'],paws['actual'])
print(f'The accuracy score for Model 1 is {acc_mod_2:.2%}')

The accuracy score for Model 1 is 63.04%


In [185]:
acc_mod_3 = accuracy_score(paws['model3'],paws['actual'])
print(f'The accuracy score for Model 1 is {acc_mod_3:.2%}')

The accuracy score for Model 1 is 50.96%


In [187]:
acc_mod_4 = accuracy_score(paws['model4'],paws['actual'])
print(f'The accuracy score for Model 1 is {acc_mod_4:.2%}')

The accuracy score for Model 1 is 74.26%


#### 4. 2. sklearn.metrics.precision_score

In [159]:
precision_score?

Signature:
precision_score(
    y_true,
    y_pred,
    *,
    labels=None,
    pos_label=1,
    average='binary',
    sample_weight=None,
    zero_division='warn',
)
Docstring:
Compute the precision.

The precision is the ratio ``tp / (tp + fp)`` where ``tp`` is the number of
true positives and ``fp`` the number of false positives. The precision is
intuitively the ability of the classifier not to label as positive a sample
that is negative.

The best value is 1 and the worst value is 0.

Read more in the :ref:`User Guide <precision_recall_f_measure_metrics>`.

Parameters
----------
y_true : 1d array-like, or label indicator array / sparse matrix
    Ground truth (correct) target values.

y_pred : 1d array-like, or label indicator array / sparse matrix
    Estimated targets as returned by a classifier.

labels : array-like, default=None
    The set of labels to include when ``average != 'binary'``, and their
    order if ``average is None``. Labels present in the data can be
    exclud

In [169]:
prec_mod_1 = precision_score(paws['actual'], paws['model1'], pos_label ='dog')
print(f'The precision score for Model 1 is {prec_mod_1:.2%}')

The precision score for Model 1 is 89.00%


In [168]:
prec_mod_2 = precision_score(paws['actual'], paws['model2'], pos_label ='dog')
print(f'The precision score for Model 1 is {prec_mod_2:.2%}')

The precision score for Model 1 is 89.32%


In [170]:
prec_mod_3 = precision_score(paws['actual'], paws['model3'], pos_label ='dog')
print(f'The precision score for Model 1 is {prec_mod_3:.2%}')

The precision score for Model 1 is 65.99%


In [171]:
prec_mod_4 = precision_score(paws['actual'], paws['model4'], pos_label ='dog')
print(f'The precision score for Model 1 is {prec_mod_4:.2%}')

The precision score for Model 1 is 73.12%


#### 4. 3. sklearn.metrics.recall_score

In [161]:
recall_score?

Signature:
recall_score(
    y_true,
    y_pred,
    *,
    labels=None,
    pos_label=1,
    average='binary',
    sample_weight=None,
    zero_division='warn',
)
Docstring:
Compute the recall.

The recall is the ratio ``tp / (tp + fn)`` where ``tp`` is the number of
true positives and ``fn`` the number of false negatives. The recall is
intuitively the ability of the classifier to find all the positive samples.

The best value is 1 and the worst value is 0.

Read more in the :ref:`User Guide <precision_recall_f_measure_metrics>`.

Parameters
----------
y_true : 1d array-like, or label indicator array / sparse matrix
    Ground truth (correct) target values.

y_pred : 1d array-like, or label indicator array / sparse matrix
    Estimated targets as returned by a classifier.

labels : array-like, default=None
    The set of labels to include when ``average != 'binary'``, and their
    order if ``average is None``. Labels present in the data can be
    excluded, for example to calculate a

In [172]:
rec_mod_1 = recall_score(paws['actual'], paws['model1'], pos_label ='dog')
print(f'The recall score for Model 1 is {rec_mod_1:.2%}')

The recall score for Model 1 is 80.33%


In [173]:
rec_mod_2 = recall_score(paws['actual'], paws['model2'], pos_label ='dog')
print(f'The recall score for Model 1 is {rec_mod_2:.2%}')

The recall score for Model 1 is 49.08%


In [174]:
rec_mod_3 = recall_score(paws['actual'], paws['model3'], pos_label ='dog')
print(f'The recall score for Model 1 is {rec_mod_3:.2%}')

The recall score for Model 1 is 50.86%


In [175]:
rec_mod_4 = recall_score(paws['actual'], paws['model4'], pos_label ='dog')
print(f'The recall score for Model 1 is {rec_mod_4:.2%}')

The recall score for Model 1 is 95.57%


#### 4. 4. sklearn.metrics.classification_report

In [158]:
classification_report?

Signature:
classification_report(
    y_true,
    y_pred,
    *,
    labels=None,
    target_names=None,
    sample_weight=None,
    digits=2,
    output_dict=False,
    zero_division='warn',
)
Docstring:
Build a text report showing the main classification metrics.

Read more in the :ref:`User Guide <classification_report>`.

Parameters
----------
y_true : 1d array-like, or label indicator array / sparse matrix
    Ground truth (correct) target values.

y_pred : 1d array-like, or label indicator array / sparse matrix
    Estimated targets as returned by a classifier.

labels : array-like of shape (n_labels,), default=None
    Optional list of label indices to include in the report.

target_names : list of str of shape (n_labels,), default=None
    Optional display names matching the labels (same order).

sample_weight : array-like of shape (n_samples,), default=None
    Sample weights.

digits : int, default=2
    Number of digits for formatting output floating point values.
    When `

In [178]:
class_mod_1 = classification_report(paws['actual'],paws['model1'], target_names = ['Cats', 'Dogs'])
print(f'The Classification report for Model 1 is:\n{class_mod_1}')

The Classification report for Model 1 is:
              precision    recall  f1-score   support

        Cats       0.69      0.82      0.75      1746
        Dogs       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [180]:
class_mod_2 = classification_report(paws['actual'],paws['model2'], target_names=['Cats', 'Dogs'])
print(f'The Classification report for Model 2 is:\n{class_mod_2}')

The Classification report for Model 2 is:
              precision    recall  f1-score   support

        Cats       0.48      0.89      0.63      1746
        Dogs       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [181]:
class_mod_3 = classification_report(paws['actual'],paws['model3'], target_names=['Cats', 'Dogs'])
print(f'The Classification report for Model 3 is:\n{class_mod_3}')

The Classification report for Model 3 is:
              precision    recall  f1-score   support

        Cats       0.36      0.51      0.42      1746
        Dogs       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [182]:
class_mod_4 = classification_report(paws['actual'],paws['model4'], target_names=['Cats', 'Dogs'])
print(f'The Classification report for Model 4 is:\n{class_mod_4}')

The Classification report for Model 4 is:
              precision    recall  f1-score   support

        Cats       0.81      0.35      0.48      1746
        Dogs       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000

